### PR TITLE
Initial Library Update for Params/Templating

### PR DESCRIPTION
**Pull request by Aric1088**
We update the generic notifiers library to support params and templating. Notifier specific implementations will be in separate pull requests.

### CODE REVIEW

File: lib/notifiers/notifiers.go [hunk 1]
```diff
@@ -322,13 +364,51 @@ func getGCSConfig(ctx context.Context, grf gcsReaderFactory, path string) (*Conf
...
+func getGCSTemplate(ctx context.Context, grf gcsReaderFactory, path string) (string, error) {
+	if trm := strings.TrimPrefix(path, "gs://"); trm != path {
+		// path started with the prefix
+		path = trm
+	} else {
+		return "", fmt.Errorf("expected %q to start with `gs://`", path)
+	}
+
+	split := strings.SplitN(path, "/", 2)
+	log.V(2).Infof("got path split: %+v", split)
```
Comment (prabenzom) [inline lib/notifiers/notifiers.go:377]: remove?

File: lib/notifiers/notifiers.go [hunk 2]
```diff
@@ -322,13 +364,51 @@ func getGCSConfig(ctx context.Context, grf gcsReaderFactory, path string) (*Conf
 	return cfg, nil
 }
 
+// getGCSConfig fetches the Template file from the given GCS path and returns the parsed Config.
+func getGCSTemplate(ctx context.Context, grf gcsReaderFactory, path string) (string, error) {
+	if trm := strings.TrimPrefix(path, "gs://"); trm != path {
+		// path started with the prefix
```
Comment (prabenzom) [inline lib/notifiers/notifiers.go:370]: it seems like you can just use a regex for this.

File: lib/notifiers/notifiers.go [hunk 3]
```diff
@@ -214,19 +237,38 @@ func Main(notifier Notifier) error {
...
+		if _, ok := allowedTemplateTypes[cfg.Spec.Notification.Template.Type]; !ok {
+			return fmt.Errorf("got invalid Template Type: %v", cfg.Spec.Notification.Template.Type)
+		}
+		if cfg.Spec.Notification.Template.URI != "" {
+			tmpl, err = getGCSTemplate(ctx, &actualGCSReaderFactory{sc}, cfg.Spec.Notification.Template.URI)
+			if err != nil {
+				return fmt.Errorf("failed to get template from GCS: %w", err)
+			}
+		} else {
+			tmpl = cfg.Spec.Notification.Template.Content
```
Comment (prabenzom) [inline lib/notifiers/notifiers.go:257]: what about a unit test for this use case?

**Review by prabenzom — CHANGES_REQUESTED** [review]

File: lib/notifiers/jsonpath.go [hunk 4]
```diff
@@ -39,10 +34,7 @@ type jpResolver struct {
 
 func newResolver(cfg *Config) (BindingResolver, error) {
 	jps := map[string]*inputAndJSONPath{}
-	for name, path := range cfg.Spec.Notification.Substitutions {
```
[the remark this answers is not in the document]
Comment (Aric1088) [reply lib/notifiers/jsonpath.go:42]: we are no longer requiring bindings/substitutions to follow a specific naming pattern

File: lib/notifiers/jsonpath.go [hunk 5]
```diff
@@ -67,22 +59,10 @@ func (j *jpResolver) Resolve(ctx context.Context, sg SecretGetter, build *cbpb.B
 	j.mtx.RLock()
 	defer j.mtx.RUnlock()
 
-	// Fetch all of the secrets that might appear in the given paths.
-	// TODO(ljr): We can try to be clever and only use the ones that are mentioned in the user-provided paths.
```
[the remark this answers is not in the document]
Comment (Aric1088) [reply lib/notifiers/jsonpath.go:71]: it was decided that we would not allow secrets available as params for the user to use in templating

File: lib/notifiers/notifiers.go (hunk 1 above)
Comment (Aric1088) [reply lib/notifiers/notifiers.go:377]: done

File: lib/notifiers/notifiers.go (hunk 2 above)
Comment (Aric1088) [reply lib/notifiers/notifiers.go:370]: done

File: lib/notifiers/notifiers.go (hunk 3 above)
Comment (Aric1088) [reply lib/notifiers/notifiers.go]: extracted the logic from that section into a testable function and added tests for it

**Review by prabenzom — APPROVED** [review]

**Pull request merged**